src/proxyvisor.js

```diff
@@ -0,0 +1,1012 @@
+		const { uuid } = req.params;
+		return db
+			.models('dependentDevice')
+			.select()
+			.where({ uuid })
+			.then(function([device]) {
+				if (device == null) {
+					return res.status(404).send('Device not found');
+				}
+				if (device.markedForDeletion) {
+					return res.status(410).send('Device deleted');
+				}
+				return res.json(parseDeviceFields(device));
+			})
+			.catch(function(err) {
+				log.error(`Error on ${req.method} ${url.parse(req.url).pathname}`, err);
+				return res.status(503).send(err?.message || err || 'Unknown error');
+			});
+	});
+
+	router.post('/v1/devices/:uuid/logs', function(req, res) {
+		const { uuid } = req.params;
+		const m = {
+			message: req.body.message,
+			timestamp: req.body.timestamp || Date.now(),
+		};
+		if (req.body.isSystem != null) {
+			m.isSystem = req.body.isSystem;
+		}
+
+		return db
+			.models('dependentDevice')
+			.select()
+			.where({ uuid })
+			.then(function([device]) {
+				if (device == null) {
+					return res.status(404).send('Device not found');
+				}
+				if (device.markedForDeletion) {
+					return res.status(410).send('Device deleted');
+				}
+				proxyvisor.logger.logDependent(m, uuid);
+				return res.status(202).send('OK');
+			})
+			.catch(function(err) {
+				log.error(`Error on ${req.method} ${url.parse(req.url).pathname}`, err);
+				return res.status(503).send(err?.message || err || 'Unknown error');
+			});
+	});
+
+	router.put('/v1/devices/:uuid', function(req, res) {
+		const { uuid } = req.params;
+		let {
+			status,
+			is_online,
+			commit,
+			releaseId,
+			environment,
+			config,
+		} = req.body;
+		const validateDeviceFields = function() {
+			if (isDefined(is_online) && !_.isBoolean(is_online)) {
+				return 'is_online must be a boolean';
+			}
+			if (!validStringOrUndefined(status)) {
+				return 'status must be a non-empty string';
+			}
+			if (!validStringOrUndefined(commit)) {
+				return 'commit must be a non-empty string';
+			}
+			if (!validStringOrUndefined(releaseId)) {
+				return 'commit must be a non-empty string';
+			}
+			if (!validObjectOrUndefined(environment)) {
+				return 'environment must be an object';
+			}
+			if (!validObjectOrUndefined(config)) {
+				return 'config must be an object';
+			}
+			return null;
+		};
+		const requestError = validateDeviceFields();
+		if (requestError != null) {
+			res.status(400).send(requestError);
+			return;
+		}
+
+		if (isDefined(environment)) {
+			environment = JSON.stringify(environment);
+		}
+		if (isDefined(config)) {
+			config = JSON.stringify(config);
+		}
+
+		const fieldsToUpdateOnDB = _.pickBy(
+			{ status, is_online, commit, releaseId, config, environment },
+			isDefined,
+		);
+		/** @type {_.Dictionary<any>} */
+		const fieldsToUpdateOnAPI = _.pick(
+			fieldsToUpdateOnDB,
+			'status',
+			'is_online',
+			'releaseId',
+		);
+		if (fieldsToUpdateOnDB.commit != null) {
+			fieldsToUpdateOnAPI.is_on__commit = fieldsToUpdateOnDB.commit;
+		}
```

Comment:
I simplified this from
```coffee
		fieldsToUpdateOnAPI = _.pick(fieldsToUpdateOnDB, 'status', 'is_online', 'commit', 'releaseId')
		if fieldsToUpdateOnAPI.commit?
			fieldsToUpdateOnAPI.is_on__commit = fieldsToUpdateOnAPI.commit
			delete fieldsToUpdateOnAPI.commit
```
whilst trying to get the types to be happy but ended up having to use `/** @type {_.Dictionary<any>} */` - I've left the simplified version though as I think it's an improvement

Reply:
We use `Dictionary<T>` everywhere already, is that accepted as a type, even though it's defined in typings/global.d.ts?

Reply:
Done